kornia/nerf/camera_utils.py

```diff
@@ -154,3 +57,67 @@ def create_spiral_path(cameras: PinholeCamera, rad: float, num_views: int, num_c
     height = torch.tensor([cameras.height[0]] * num_views, device=device)
     width = torch.tensor([cameras.width[0]] * num_views, device=device)
     return PinholeCamera(mean_intrinsics, extrinsics, height, width)
+
+
+def create_spline_curve(cameras: PinholeCamera, num_views: int) -> PinholeCamera:
+    r"""Create a PinholeCamera object with cameras that are positioned along a path that is calculated from the
+    locations of the input camera centers. Parameteric spline is used for the tranjectory. Output camera intrinsics
+    are also interpolated from those of the input cameras.
+
+    Args:
+        cameras: Scene cameras used to train the NeRF model: PinholeCamera
+        num_views: Number of created cameras: int
+    """
+    # Extrinsics
+    origins = cameras.origins().cpu().squeeze().numpy()
+    ox = origins[:, 0]
+    oy = origins[:, 1]
+    oz = origins[:, 2]
+
+    rotation_matrix = torch.clone(cameras.rotation_matrix).cpu()  # Cloning to restore memory contingency
+    q = rotation_matrix_to_quaternion(rotation_matrix, order=QuaternionCoeffOrder.WXYZ).numpy()
+    qw = q[:, 0]
+    qx = q[:, 1]
+    qy = q[:, 2]
+    qz = q[:, 3]
+
+    # Intrinsics
+    fx_new = torch.mean(cameras.fx).item()  # Use average of camera focals
+    fy_new = torch.mean(cameras.fy).item()
+    width_new = cameras.width[0].item()  # For new camera sizes use first input camera
+    height_new = cameras.height[0].item()
+    cx_new = width_new / 2.0  # Assign principal point to mid-image
+    cy_new = height_new / 2.0
+
+    # Interpolate output camera extrinsic parameters
+    tck, _ = splprep([ox, oy, oz, qw, qx, qy, qz], s=0)
+    u2 = torch.linspace(0, 1, num_views, device='cpu', dtype=cameras.dtype).numpy()
+    new_points = splev(u2, tck)
+
+    device = cameras.device
+    dtype = cameras.dtype
+    intrinsics: List[Tensor] = []
+    extrinsics: List[Tensor] = []
+    for i in range(num_views):
+        intrinsic = torch.eye(4, device=device, dtype=dtype)
+        intrinsic[0, 0] = fx_new
+        intrinsic[1, 1] = fy_new
+        intrinsic[0, 2] = cx_new
+        intrinsic[1, 2] = cy_new
+        intrinsics.append(intrinsic)
+
+        q_new = torch.tensor([new_points[3][i], new_points[4][i], new_points[5][i], new_points[6][i]], device=device)
+        R_new = quaternion_to_rotation_matrix(q_new, order=QuaternionCoeffOrder.WXYZ)
+        o_new = torch.tensor([new_points[0][i], new_points[1][i], new_points[2][i]], device=device)
```

Comment:
i guess here we can use too `Se3` and project points

Reply:
OK. For now I am removing the spline based path planning function